sam_learning/core/propositional_operations/literals_cnf.py

```diff
@@ -67,6 +67,35 @@ def is_action_safe(self, action_name: str, action_preconditions: Set[str]) -> bo
 
         return True
 
+    def is_action_safe(self, action_name: str, action_preconditions: Set[str]) -> bool:
+        """Checks if an action is safe to execute based on this CNF clause.
+
+               :param action_name: the name of the action.
+               :param action_preconditions: the preconditions of the action.
+               :return: True if the action is safe to execute, False otherwise.
+               """
+
+        unitc = []  # unit clauses of size 1 (1 tuple) that contains action_name
+        nonunc = []  # non-unit clauses that contain action_name
+
+        for lifted_options in self.possible_lifted_effects:
+            if action_name in [action for (action, _) in lifted_options]:
+                if len(lifted_options) == 1:
+                    unitc.append(lifted_options[0])
+                else:
+                    nonunc.append(lifted_options)
+
+        if len(nonunc) == 0:
+            return True
+
+        for nunc in nonunc:
+            if not any(uc in nunc for uc in unitc):
```

Comment:
this works?

Reply:
let me better check that with some tests as you suggested